types/wlr_wl_shell.c

```diff
@@ -1,61 +1,264 @@
 #include <wayland-server.h>
 #include <wlr/util/log.h>
+#include <wlr/types/wlr_surface.h>
 #include <wlr/types/wlr_wl_shell.h>
 #include <stdlib.h>
+#include <wayland-server-protocol.h>
 
-static void shell_surface_pong(struct wl_client *client, struct wl_resource
-		*resource, uint32_t serial) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface pong");
+static void shell_surface_pong(struct wl_client *client,
+		struct wl_resource *resource, uint32_t serial) {
+	wlr_log(L_DEBUG, "got shell surface pong");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+	if (surface->ping_serial != serial) {
+		return;
+	}
+
+	wl_event_source_timer_update(surface->ping_timer, 0);
+	surface->ping_serial = 0;
 }
 
-static void shell_surface_move(struct wl_client *client, struct wl_resource
-		*resource, struct wl_resource *seat, uint32_t serial) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface move");
+static void shell_surface_move(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *seat_resource,
+		uint32_t serial) {
+	wlr_log(L_DEBUG, "got shell surface move");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+	struct wlr_seat_handle *seat_handle =
+		wl_resource_get_user_data(seat_resource);
+
+	struct wlr_wl_shell_surface_move_event *event =
+		calloc(1, sizeof(struct wlr_wl_shell_surface_move_event));
+	if (event == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	event->client = client;
+	event->surface = surface;
+	event->seat_handle = seat_handle;
+	event->serial = serial;
+
+	wl_signal_emit(&surface->events.request_move, event);
+
+	free(event);
 }
 
 static void shell_surface_resize(struct wl_client *client,
-		struct wl_resource *resource, struct wl_resource *seat, uint32_t serial,
-		uint32_t edges) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface resize");
+		struct wl_resource *resource, struct wl_resource *seat_resource,
+		uint32_t serial, uint32_t edges) {
+	wlr_log(L_DEBUG, "got shell surface resize");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+	struct wlr_seat_handle *seat_handle =
+		wl_resource_get_user_data(seat_resource);
+
+	struct wlr_wl_shell_surface_resize_event *event =
+		calloc(1, sizeof(struct wlr_wl_shell_surface_resize_event));
+	if (event == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	event->client = client;
+	event->surface = surface;
+	event->seat_handle = seat_handle;
+	event->serial = serial;
+	event->edges = edges;
+
+	wl_signal_emit(&surface->events.request_resize, event);
+
+	free(event);
 }
 
 static void shell_surface_set_toplevel(struct wl_client *client,
 		struct wl_resource *resource) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface set_toplevel");
+	wlr_log(L_DEBUG, "got shell surface toplevel");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+
+	if (surface->role != WLR_WL_SHELL_SURFACE_ROLE_NONE) {
```

Comment:
You should be able to set a surface that was previously in another state to a toplevel. It should undo the previous state things like the parent and all that.

Reply:
Okay. I suppose it's also possible to transform a toplevel to a popup/transcient?